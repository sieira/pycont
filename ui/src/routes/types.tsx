import * as React from 'react';

export interface AuthStateRouteProps {
  exact?: boolean;
  isAuthenticated: boolean | null;
  path: string;
  component: React.ComponentType<any>;
}
