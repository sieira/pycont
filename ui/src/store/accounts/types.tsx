export interface Account {
  id: number
  name?: string
  balance?: number
}

export interface AccountsState {
  accountList: Account[]
  fetched: boolean
}
