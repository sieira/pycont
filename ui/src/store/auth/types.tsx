export interface AuthState {
  isAuthenticated: boolean | null
  csrfToken: string | null
}
