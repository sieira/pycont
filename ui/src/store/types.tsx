import { AuthState } from 'auth/types'
import { AccountsState } from 'accounts/types'

export interface PycontState {
  auth: AuthState
  accounts: AccountsState
}
