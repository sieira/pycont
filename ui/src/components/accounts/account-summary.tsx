import React from 'react'
import { connect } from 'react-redux'
import { Account } from '../../store/accounts/types'
import { Spinner, Form } from 'react-bootstrap'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Button } from 'react-bootstrap'

export default class AccountSummary extends React.Component<StateProps> {
  constructor(props) {
    super(props)

    this.state = {
      isEditing: false,
      name: this.props.account.name,
      balance: this.props.account.balance
    }

    this.toggleEdit = this.toggleEdit.bind(this)
  }

  toggleEdit() {
    this.setState(prevState => ({ isEditing: !prevState.isEditing }))
  }

  render() {
    return this.state.isEditing ? (
      <Form>
        <Form.Row>
          <Form.Group>
            <Form.Control
              type="text"
              value={this.state.name}
              onChange={e => this.setState({ name: e.target.value })}
            />
          </Form.Group>
          <Form.Group>
            <Form.Control
              type="text"
              value={this.state.balance}
              onChange={e => this.setState({ balance: e.target.value })}
            />
          </Form.Group>
          <Button onClick={this.toggleEdit}>
            <FontAwesomeIcon icon="check-square" />
          </Button>
          <Button onClick={this.toggleEdit}>
            <FontAwesomeIcon icon="times" />
          </Button>
        </Form.Row>
      </Form>
    ) : (
      <>
        <span>{this.props.account.name}</span>
        <span>{this.props.account.balance}</span>
        <Button onClick={this.toggleEdit}>
          <FontAwesomeIcon icon="pencil-alt" />
        </Button>
      </>
    )
  }
}
