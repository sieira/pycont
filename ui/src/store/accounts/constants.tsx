export const FETCH = 'FETCH'
export type FETCH = typeof FETCH
