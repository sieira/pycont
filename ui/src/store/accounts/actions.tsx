import Cookies from 'universal-cookie'

import { Action } from 'redux'
import { ThunkDispatch as Dispatch } from 'redux-thunk'

import * as constants from './constants'
import { Account, AccountsState } from './types'

export interface Fetch {
  type: constants.FETCH
  payload: AccountsState
}

export interface Patch {
  type: constants.PATCH
  payload: AccountsState
}

export function fetchData(accounts: Account[]): Fetch {
  return {
    type: constants.FETCH,
    payload: { accountList: accounts, fetched: true }
  }
}

export function resetData(): Patch {
  return {
    type: constants.PATCH,
    payload: { accountList: [], fetched: false }
  }
}

export type AccountsAction = Fetch | Patch

export function fetchAccounts() {
  return async (
    dispatch: Dispatch<AccountsAction, {}, Action>
  ): Promise<void> => {
    return fetch('api/accounts/', {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json'
      }
    }).then(function(response) {
      response.json().then(data => {
        dispatch(fetchData(data))
      })
    })
  }
}

export function patchAccount(account) {
  return async (
    dispatch: Dispatch<AccountsAction, {}, Action>
  ): Promise<void> => {
    return fetch(`api/accounts/${account.id}/`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        'X-CSRFToken': new Cookies().get('csrftoken')
      },
      body: JSON.stringify(account)
    }).then(function(response) {
      response.json().then(data => {
        dispatch(resetData())
        dispatch(fetchAccounts())
      })
    })
  }
}
