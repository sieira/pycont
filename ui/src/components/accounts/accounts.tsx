import React from 'react'
import { connect } from 'react-redux'
import { ThunkDispatch as Dispatch } from 'redux-thunk'

import { Spinner, ListGroup } from 'react-bootstrap'

import { PycontState } from '../../store/types'
import { Account } from '../../store/accounts/types'
import { fetchAccounts } from '../../store/accounts/actions'

interface StateProps {
  accounts: Account[]
  fetched: boolean
}

interface DispatchProps {
  fetchData: () => void
}

class AccountList extends React.Component<StateProps & DispatchProps> {
  componentDidMount(): void {
    if (!this.props.fetched) {
      this.props.fetchData()
    }
  }

  render() {
    return (
      <>
        <h2>Accounts</h2>
        <div aria-busy={!this.props.fetched} aria-live="polite">
          {!this.props.fetched ? (
            <Spinner animation="grow" variant="info" />
          ) : (
            <ListGroup>
              {this.props.accounts.map(account => (
                <ListGroup.Item key={account.id}>{account.name}</ListGroup.Item>
              ))}
            </ListGroup>
          )}
        </div>
      </>
    )
  }
}

const mapStateToProps = (state: PycontState): StateProps => ({
  accounts: state.accounts.accountList,
  fetched: state.accounts.fetched
})

const mapDispatchToProps = (
  dispatch: Dispatch<PycontState, any, any>
): DispatchProps => ({
  fetchData: () => dispatch(fetchAccounts())
})

export default connect(mapStateToProps, mapDispatchToProps)(AccountList)
