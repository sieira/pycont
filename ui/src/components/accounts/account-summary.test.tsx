import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk-recursion-detect'

import { mount } from 'enzyme'

import { Account } from '../../store/accounts/types'
import AccountSummary from './account-summary'

import { createWithProvider } from '../../test-utils/render-utils.test'

const mockStore = configureMockStore([thunkMiddleware])
const store = mockStore({ auth: { isAuthenticated: false } })

it('should render without crashing', () => {
  const accountSummary = createWithProvider(
    <AccountSummary account={{ name: 'SomeAccount', balance: 200 }} />
  )
  expect(accountSummary.toJSON()).toMatchSnapshot()
})

it('should toggle isEditing when click on edit', () => {
  const wrapper = mount(
    <Provider store={store}>
      <AccountSummary account={{ name: 'SomeAccount', balance: 200 }} />
    </Provider>
  )
  const accountSummary = wrapper.find('AccountSummary')
  expect(accountSummary.state('isEditing')).toBe(false)
  const editButton = accountSummary.find('button')
  expect(editButton.props().disabled).toBe(false)
  editButton.simulate('click')
  expect(accountSummary.state('isEditing')).toBe(true)
})
