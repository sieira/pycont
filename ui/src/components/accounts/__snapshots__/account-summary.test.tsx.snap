// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
Array [
  <span>
    SomeAccount
  </span>,
  <span>
    200
  </span>,
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  />,
]
`;
